File: web/src/api/accounting.ts
import { apiFetch } from './client'

export interface AccountingSummary {
  wallet_id: string
  method: string
  fiat_currency: string
  total_cost_sats: number
  total_cost_fiat: number
  unrealised_gain_fiat: number
  realised_gain_fiat: number
  computed_at: string
}

export interface CostBasisRecord {
  id: string
  wallet_id: string
  txid: string
  vout: number
  acquired_at: string
  cost_sats: number
  cost_fiat: number
  fiat_currency: string
  method: string
  disposed_at: string | null
  proceeds_fiat: number | null
  gain_fiat: number | null
}

export type AccountingMethod = 'fifo' | 'avgcost' | 'lifo' | 'hifo' | 'specificid' | 'section104'

export const runAccounting = (
  walletID: string,
  method: AccountingMethod,
  currency: string
) =>
  apiFetch<AccountingSummary>(`/wallets/${walletID}/accounting/run`, {
    method: 'POST',
    body: JSON.stringify({ method, currency }),
  })

export const getAccountingSummary = (walletID: string) =>
  apiFetch<AccountingSummary>(`/wallets/${walletID}/accounting/summary`)

export const listCostBasis = (walletID: string) =>
  apiFetch<CostBasisRecord[]>(`/wallets/${walletID}/accounting/cost-basis`)

File: internal/domain/accounting.go
package domain

import "time"

type CostBasisMethod string

const (
	MethodFIFO       CostBasisMethod = "fifo"
	MethodAvgCost    CostBasisMethod = "avgcost"
	MethodLIFO       CostBasisMethod = "lifo"
	MethodHIFO       CostBasisMethod = "hifo"
	MethodSpecificID  CostBasisMethod = "specificid"
	MethodSection104  CostBasisMethod = "section104"
)

type CostBasisRecord struct {
	ID           string          `json:"id"`
	WalletID     string          `json:"wallet_id"`
	Txid         string          `json:"txid"`
	Vout         int             `json:"vout"`
	AcquiredAt   time.Time       `json:"acquired_at"`
	CostSats     int64           `json:"cost_sats"`
	CostFiat     int64           `json:"cost_fiat"`   // cents
	FiatCurrency string          `json:"fiat_currency"`
	Method       CostBasisMethod `json:"method"`
	DisposedAt   *time.Time      `json:"disposed_at,omitempty"`
	ProceedsFiat *int64          `json:"proceeds_fiat,omitempty"` // cents
	GainFiat     *int64          `json:"gain_fiat,omitempty"`     // cents
}

type PriceSnapshot struct {
	ID         string    `json:"id"`
	Currency   string    `json:"currency"`
	PriceFiat  int64     `json:"price_fiat"` // price per BTC in cents
	Source     string    `json:"source"`     // e.g. "manual", "coingecko"
	Timestamp  time.Time `json:"timestamp"`
}
